contrib/win32/win32compat/ssh-agent/authagent-request.c

```diff
@@ -250,6 +252,128 @@ duplicate_token_for_client(struct agent_connection* con, HANDLE t) {
+	LSA_OPERATIONAL_MODE mode;
+	ULONG auth_package_id, logon_info_size = 0;
+	NTSTATUS ret, subStatus;
+	wchar_t *logon_info = NULL;
+	LSA_STRING logon_process_name, lsa_auth_package_name, originName;
+	TOKEN_SOURCE sourceContext;
+	PVOID pProfile = NULL;
+	LUID logonId;
+	QUOTA_LIMITS quotas;
+	DWORD cbProfile;
+	int retVal = -1;
+
+	if (sshbuf_get_string_direct(request, &user, &user_len) != 0 ||
+		sshbuf_get_cstring(request, &pwd, &pwd_len) != 0 ||
+		sshbuf_get_string_direct(request, &domain, &domain_len) != 0 ||
+		user_len > MAX_USER_LEN || pwd_len == 0 ||
+		sshbuf_get_string_direct(request, &lsa_pkg, &lsa_pkg_len) != 0) {
+		debug("invalid LSA auth request");
+		goto done;
+	}
+
+	debug("LSA auth request, user:%s domain:%s lsa_pkg:%s ", user, domain, lsa_pkg);
+
+	/* convert everything to utf16 only if its not NULL */
+	if ((userw = utf8_to_utf16(user)) == NULL ||
+		(pwdw = utf8_to_utf16(pwd)) == NULL ||
+		(domain && (domw = utf8_to_utf16(domain)) == NULL))	{
+		debug("%s: out of memory", __func__);
+		goto done;
+	}
+
+	/* call into LSA provider , get and duplicate token */
+	InitLsaString(&logon_process_name, "ssh-agent");
+	InitLsaString(&lsa_auth_package_name, lsa_pkg);
+	InitLsaString(&originName, "sshd");
+
+	if ((ret = LsaRegisterLogonProcess(&logon_process_name, &lsa_handle, &mode)) != STATUS_SUCCESS) {
+		error("LsaRegisterLogonProcess failed, error:%ld", LsaNtStatusToWinError(ret));
+		goto done;
+	}
+
+	if ((ret = LsaLookupAuthenticationPackage(lsa_handle, &lsa_auth_package_name, &auth_package_id)) != STATUS_SUCCESS) {
+		error("LsaLookupAuthenticationPackage failed, lsa auth pkg:%ls error:%ld", lsa_pkg, LsaNtStatusToWinError(ret));
+		goto done;
+	}
+
+	logon_info_size = (ULONG)((wcslen(userw) + wcslen(pwdw) + wcslen(domw) + 3) * sizeof(wchar_t));
+	logon_info = (wchar_t *)malloc(logon_info_size);
+	if (NULL == logon_info)
+		fatal("%s:out of memory", __func__);
+
+	wcscpy_s(logon_info, logon_info_size, userw);
+	wcscat_s(logon_info, logon_info_size, L";");
+	wcscat_s(logon_info, logon_info_size, pwdw);
+	wcscat_s(logon_info, logon_info_size, L";");
+	wcscat_s(logon_info, logon_info_size, domw);
+
+	memcpy(sourceContext.SourceName, "sshd", sizeof(sourceContext.SourceName));
+
+	if (!AllocateLocallyUniqueId(&sourceContext.SourceIdentifier)) {
+		error("AllocateLocallyUniqueId failed, error:%d", GetLastError());
+		goto done;
+	}		
+
+	if ((ret = LsaLogonUser(lsa_handle,
+		&originName,
+		Network,
+		auth_package_id,
+		logon_info,
+		logon_info_size,
+		NULL,
+		&sourceContext,
+		&pProfile,
+		&cbProfile,
+		&logonId,
+		&token,
+		&quotas,
+		&subStatus)) != STATUS_SUCCESS) {		
+		if(ret == STATUS_ACCOUNT_RESTRICTION)
+			error("LsaLogonUser failed, error:%ld subStatus:%ld", LsaNtStatusToWinError(ret), subStatus);
+		else
+			error("LsaLogonUser failed error:%ld", LsaNtStatusToWinError(ret));
+
+		goto done;
+	}
+
+	if ((dup_token = duplicate_token_for_client(con, token)) == NULL)

```

Comment:
if we are sure finishing useing dup_token, better to close it here

Reply:
duplicate token is sent to the sshd.. 
closing the token here doesn't have any impact as the duplicate token is registered under client/sshd process..